sdk/servicebus/service-bus/src/track2/constructorHelpers.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Session, QueueAuth, SubscriptionAuth } from "./models";
+import { ServiceBusClientReceiverOptions } from "../old/oldServiceBusReceiverClient";
+import { ReceiveMode } from "../serviceBusMessage";
+import { isTokenCredential, TokenCredential } from '@azure/core-amqp';
+import { ConnectionContext } from '../connectionContext';
+import { createConnectionContextForTokenCredential, createConnectionContextForConnectionString, ServiceBusClientOptions } from '../old/serviceBusClient';
+
+// The methods in this file are all here just to make handling
+// the _large_ number of constructor overloads for ReceiverClient.
+//
+// As a result you'll see that some of the type-overloaded-craziness
+// bleeds into these methods. They're not pretty, but at least they're
+// factored out.
+
+/**
+ * Checks the receiveMode parameter to see if it's an actual receiveMode
+ * or one of the multitude of constructor options available to the user.
+ *
+ * @param possibleReceiveMode
+ * @internal
+ * @ignore
+ */
+export function isReceiveMode(
+  possibleReceiveMode:
+    | "peekLock"
+    | "receiveAndDelete"
+    | Session
+    | ServiceBusClientReceiverOptions
+    | undefined
+): possibleReceiveMode is "peekLock" | "receiveAndDelete" {
+  return (
+    possibleReceiveMode != null &&
+    typeof possibleReceiveMode === "string" &&
+    (possibleReceiveMode === "peekLock" || possibleReceiveMode === "receiveAndDelete")
+  );
+}
+
+/**
+ * Attempts to generically figure out what the entity path is from the grab bag of string parameters
+ * we get in our constructors.
+ *
+ * @param connectionString The connection string - could be entity specific (topic, queue) or for the service bus itself.
+ * @param optionalQueueOrSubscriptionOrTopicName An optional entity name.
+ * There are 3 different options from here:
+ * - They passed a service bus connection string in `connectionString`: this parameter will be a topic or queue
+ * - They passed an entity specific connection string (topic/queue): this parameter will be undefined _or_ a subscription
+ * @param optionalSubscriptionName An optional subscription name.
+ * You get here if the user is targeting a subscription but passed:
+ * 1. A service bus only connection string for parameter 1
+ * 2. A topic name for parameter 2
+ * 3. And now a subscription for parameter 3
+ *
+ * @internal
+ * @ignore
+ */
+export function createConnectionContext(
+  auth: QueueAuth | SubscriptionAuth,
+  options: ServiceBusClientOptions
+): { context: ConnectionContext, entityPath: string } {
+
+  // TODO: replace with actual typeguards
+  const auth2 = auth as any;
```

Comment:
Slated to be replaced with actual type guards and checking. General structure is correct but needs to be tested.

Reply:
Guards added and it cleaned things up quite nicely.

There's still some redundant code that I can refactor with but we can do that in a later PR.